sharing_portal/views.py

```diff
@@ -1,226 +1,180 @@
 from django.core.mail import send_mail
 from django.urls import reverse
-from django.db.models import F, Q, IntegerField, Count
 from django.http import Http404, HttpResponse, HttpResponseRedirect
 from django.core.exceptions import PermissionDenied
-from django.shortcuts import get_object_or_404, redirect
+from django.shortcuts import redirect
 from django.template import loader
 from django.utils.html import strip_tags
 from django.utils import timezone
 from projects.models import Project
-from allocations.models import Allocation
-from rest_framework.renderers import JSONRenderer
-from rest_framework import serializers
+from projects.util import get_project_members
 from util.project_allocation_mapper import ProjectAllocationMapper
 from util.keycloak_client import KeycloakClient
 
 from .forms import (
     ArtifactForm,
-    ArtifactVersionForm,
     AuthorFormset,
     ShareArtifactForm,
-    ZenodoPublishFormset,
     RequestDaypassForm,
     ReviewDaypassForm,
 )
-from .models import Artifact, ArtifactVersion, Author, ShareTarget, DaypassRequest
-from .tasks import publish_to_zenodo
+from .models import DaypassRequest, DaypassProject
+from . import trovi
 from projects.views import (
     add_project_invitation,
     get_invite_url,
     get_project_membership_managers,
     is_membership_manager,
     manage_membership_in_scope,
 )
+from .zenodo import ZenodoClient
+
+from urllib.parse import urlencode
 
 import logging
+from datetime import datetime, timedelta
 
 LOG = logging.getLogger(__name__)
 
 SHARING_KEY_PARAM = "s"
 
 
-class ArtifactSerializer(serializers.ModelSerializer):
-    class Meta:
-        model = Artifact
-        fields = "__all__"
+def ensure_trovi_token(view_func):
+    def _wrapped_view(request, *args, **kwargs):
+        if (
+            request.session.get("trovi_token_expiration")
+            and datetime.now().timestamp() > request.session["trovi_token_expiration"]
+        ):
+            del request.session["trovi_token_expiration"]
+            del request.session["trovi_token"]
+
+        if not request.session.get("trovi_token"):
+            if request.session.get("oidc_access_token"):
+                try:
+                    response = trovi.get_token(
+                        request.session.get("oidc_access_token"),
+                        is_admin=request.user.is_staff,
+                    )
+                    request.session["trovi_token"] = response["access_token"]
+                    request.session["trovi_token_expiration"] = min(
+                        (
+                            datetime.now() + timedelta(seconds=response["expires_in"])
+                        ).timestamp(),
+                        request.session["oidc_token_expiration"],
+                    )
+                except trovi.TroviException:
+                    LOG.error("Error getting trovi token")
+            else:
+                # Set an empty token
+                request.session["trovi_token"] = ""
```

Comment:
What is the effect of an empty token? The function implies it "ensures" a token, but it looks like all that is ensured is that the key is set on the session. Does it make sense to `raise`, or does that introduce too many hard-to-manage exception cases?

Reply:
I guess what I'm saying is, I thought from the name of the function that it would raise if it could not ensure the token was there. It's fine if it's cleaner to not do that, but maybe just change the name to `with_trovi_token` or something

Reply:
Setting the token empty is set so the unauthenticated user has no token when talking to trovi.

I've changed this to `with_trovi_token` to make this clearer.